graph/path/yen_ksp.go

```diff
@@ -0,0 +1,156 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package path
+
+import (
+	"sort"
+
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/internal/set"
+)
+
+func YenKShortestPath(g graph.Graph, k int, s, t graph.Node) [][]graph.Node {
+	yk := yenKSPAdjuster{
+		g:       g,
+		visited: make(map[int64]set.Int64s),
+	}
+
+	var weight Weighting
+	if wg, ok := g.(Weighted); ok {
+		weight = wg.Weight
+	} else {
+		weight = UniformCost(g)
+	}
+
+	yk.weight = weight
+
+	var paths [][]graph.Node
+
+	shortest, _ := DijkstraFrom(s, yk).To(t.ID())
+	if len(shortest) != 0 {
+		paths = append(paths, shortest)
+	} else {
+		return paths
+	}
+
+	var pot []yenShortest
+
+	for i := int64(1); i < int64(k); i++ {
+		for n := 0; n < (len(paths[i-1]) - 1); n++ {
+			spur := paths[i-1][n]
+			root := make([]graph.Node, len(paths[i-1][:n+1]))
+			copy(root, paths[i-1][:n+1])
+
+			var rootWeight float64
+			for x := 1; x < len(root); x++ {
+				w, _ := weight(root[x-1].ID(), root[x].ID())
+				rootWeight += w
+			}
+
+			for _, path := range paths {
+				if len(path) > n {
+					ok := true
+					for x := 0; x < len(root); x++ {
+						if path[x].ID() != root[x].ID() {
+							ok = false
+							break
+						}
+					}
+					if ok {
+						yk.addVisited(path[n].ID(), path[n+1].ID())
+					}
+				}
+			}
+
+			spath, weight := DijkstraFrom(spur, yk).To(t.ID())
+			size := len(root) - 1
+
+			if len(root) > 1 {
+				nroot := root[:size]
+				nroot = append(nroot, spath...)
+				potential := yenShortest{nroot, weight + rootWeight}
+				pot = append(pot, potential)
+			} else {
+				potential := yenShortest{spath, weight}
+				pot = append(pot, potential)
+			}
+
+			yk.visited = make(map[int64]set.Int64s)
```

Comment:
I don't believe this is correct; the algorithm removes edges from G over the life of the algorithm, while you are removing nodes over the life of a single iteration. This is completely difference and, without rigorous analysis to back this up, likely to result in re-walking paths in adequately small graphs. At least one of the tests I suggested should fail because of this.

Reply:
After reading through the wikipedia page, the way I understand it is that the algorithm is removing edges over the life of each spur path the algorithm is checking so after it has finished iterating through each spur path it resets the graph. [Wikipedia Link](https://en.wikipedia.org/wiki/Yen%27s_algorithm)

Reply:
Sorry, you're right; I missed the section below under "// Add back the edges and nodes that were removed from the graph."

Reply:
However, you need a set of edges, not a set of nodes.

Reply:
I am technically storing edges because I only use the visited map in the From function which is looking for edges. Also, the map key is the From node and all the nodes it has already visited in the past.

Reply:
Again, sorry. You are right - I missed it because it's not a common pattern here.

I will ask that you use a `map[[2]int64]struct{}` or `map[struct{ f, t int64 ]struct{}` though since that requires a single look-up rather than the two here. Note that the edges stored change depending on whether the input `g` is directed or undirected (both directions need to be stored in the undirected case).